### PR TITLE
add-basic-targets

### PR DESCRIPTION
**Pull request by omegas27**
Towards https://github.com/giantswarm/giantswarm/issues/11720

### CODE REVIEW

File: files/templates/additional-scrape-configs.template [hunk 1]
```diff
@@ -1,15 +1,188 @@
-- job_name: "docker"
+# Additional scrape configuration for the tenant clusters components
+# Add scrape configuration for docker 
+- job_name: docker
```
Comment (JosephSalisbury) [inline files/templates/additional-scrape-configs.template:3]: Can we align the job names w/ the job names generated by the prometheus-operator? Being consistent here would be neat

File: files/templates/additional-scrape-configs.template [hunk 2]
```diff
@@ -1,15 +1,188 @@
...
+  # Add cluster_id label.
+  - target_label: cluster_id
+    replacement: <CLUSTER_ID>
+  # Add cluster_type label.
+  - target_label: cluster_type
+    replacement: tenant_cluster
+  # Add provider label.
+  - target_label: provider
+    replacement: <PROVIDER>
+- job_name: aws-node
```
Comment (JosephSalisbury) [inline files/templates/additional-scrape-configs.template:146]: Only deploying `aws-node` for AWS would be neat too

File: files/templates/additional-scrape-configs.template (hunk 1 above)
Comment (omegas27) [reply files/templates/additional-scrape-configs.template:3]: Sure :)

File: files/templates/additional-scrape-configs.template (hunk 2 above)
Comment (omegas27) [reply files/templates/additional-scrape-configs.template:146]: I need to check how but yes :) This would mean a better templating tool but it should be doable

Comment (omegas27) [reply files/templates/additional-scrape-configs.template]: This is fixed :)

**Comment by omegas27** [comment]
Tested on geckon, this works but scraping is not working on Azure and AWS and I am unsure why:

`n=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:31:18.278Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:262: Failed to list *v1.Service: Get https://master.3w8xa/api/v1/services?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:31:18.278Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:261: Failed to list *v1.Endpoints: Get https://master.3w8xa/api/v1/endpoints?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:31:18.278Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:261: Failed to list *v1.Endpoints: Get https://master.3w8xa/api/v1/namespaces/kube-system/endpoints?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=info ts=2020-08-06T09:31:37.534Z caller=main.go:747 msg="Loading configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=info ts=2020-08-06T09:31:37.539Z caller=main.go:775 msg="Completed loading of configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:333: Failed to list *v1.Node: Get https://master.3w8xa/api/v1/nodes?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:261: Failed to list *v1.Endpoints: Get https://master.3w8xa/api/v1/endpoints?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:263: Failed to list *v1.Pod: Get https://master.3w8xa/api/v1/pods?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:261: Failed to list *v1.Endpoints: Get https://master.3w8xa/api/v1/namespaces/kube-system/endpoints?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:262: Failed to list *v1.Service: Get https://master.3w8xa/api/v1/namespaces/kube-system/services?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:281: Failed to list *v1.Pod: Get https://master.3w8xa/api/v1/pods?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.478Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:262: Failed to list *v1.Service: Get https://master.3w8xa/api/v1/services?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"
level=error ts=2020-08-06T09:32:09.479Z caller=klog.go:94 component=k8s_client_runtime func=ErrorDepth msg="/app/discovery/kubernetes/kubernetes.go:263: Failed to list *v1.Pod: Get https://master.3w8xa/api/v1/namespaces/kube-system/pods?limit=500&resourceVersion=0: dial tcp 172.31.141.168:443: connect: connection timed out"


Also, we cannot scrape nginx for 12.0.0 clusters
`

**Comment by omegas27** [comment]
Tested on geckon and godsmack

**Comment by omegas27** [comment]
ping @giantswarm/team-atlas

File: service/controller/resource/prometheus/resource.go [hunk 3]
```diff
@@ -73,6 +73,10 @@ func toPrometheus(v interface{}) (metav1.Object, error) {
 					KeyFile:  fmt.Sprintf("/etc/prometheus/secrets/%s/key", key.Secret()),
 				},
 			},
+			ExternalLabels: map[string]string{
```
Comment (mig4) [inline service/controller/resource/prometheus/resource.go:76]: What's this for?
Also, do you not need to add reference to your `additionalScrapeConfigs` secret here?

File: service/controller/resource/servicemonitor/service/apiserver.go [hunk 4]
```diff
@@ -28,9 +28,46 @@ func APIServer(cluster metav1.Object) *promv1.ServiceMonitor {
...
+						{
+							TargetLabel: "cluster_type",
+							Replacement: "tenant_cluster",
+						},
+						{
+							TargetLabel: "provider",
+							Replacement: provider,
+						},
+						{
+							SourceLabels: []string{"__meta_kubernetes_service_annotation_giantswarm_io_monitoring"},
```
Comment (mig4) [inline service/controller/resource/servicemonitor/service/apiserver.go:66]:
```suggestion
							SourceLabels: []string{"__meta_kubernetes_service_label_giantswarm_io_monitoring"},
```

File: service/controller/resource/servicemonitor/service/cluster_autoscaler.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+package service
```
Comment (mig4) [inline service/controller/resource/servicemonitor/service/cluster_autoscaler.go:1]: It'd be better to have this in a separate PR as this one is quite big and appears to cover a number of things making it a bit hard to review. See also [our docs on PRs](https://intranet.giantswarm.io/docs/dev-and-releng/developer-workflow/pull-requests/). It's more of a note for the future, probably not worth re-working this one. Not sure how inter-dependent those changes are.

File: pkg/templates/templates.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	"text/template"
+
+	"github.com/giantswarm/microerror"
+)
+
+func Render(templates []string, data interface{}) (string, error) {
+	var err error
+
+	main := template.New("main")
+	for _, t := range templates {
```
Comment (mig4) [inline pkg/templates/templates.go:14]: What's the use-case for a list of templates here? Isn't it that last one wins (overrides prior ones) anyway?

File: files/templates/additional-scrape-configs.template.yaml [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  # Add cluster_id label.
+  - target_label: cluster_id
+    replacement: {{ .ClusterID }}
+  # Add cluster_type label.
+  - target_label: cluster_type
+    replacement: tenant_cluster
+  # Add provider label.
+  - target_label: provider
+    replacement: {{ .Provider }}
+{{ end }}
```
Comment (mig4) [inline files/templates/additional-scrape-configs.template.yaml:190]: nit: add newline 😁

File: service/controller/resource/servicemonitor/service/cluster_autoscaler.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+						{
+							TargetLabel: "cluster_type",
+							Replacement: "tenant_cluster",
+						},
+						{
+							TargetLabel: "provider",
+							Replacement: provider,
+						},
+						{
+							SourceLabels: []string{"__meta_kubernetes_service_annotation_giantswarm_io_monitoring"},
```
Comment (mig4) [inline service/controller/resource/servicemonitor/service/cluster_autoscaler.go:76]:
```suggestion
							SourceLabels: []string{"__meta_kubernetes_service_label_giantswarm_io_monitoring"},
```

File: service/controller/resource/servicemonitor/service/nginx-ingress-controller.go [hunk 9]
```diff
@@ -28,22 +28,55 @@ func NginxIngressController(cluster metav1.Object) *promv1.ServiceMonitor {
 				MatchNames: []string{"kube-system"},
 			},
 			Endpoints: []promv1.Endpoint{
-				promv1.Endpoint{
-					Port:   "https",
-					Scheme: "https",
+				{
```
Comment (mig4) [inline service/controller/resource/servicemonitor/service/nginx-ingress-controller.go:31]: Are these changes related to the others in this PR?

File: service/controller/resource/servicemonitor/service/nginx-ingress-controller.go [hunk 10]
```diff
@@ -28,22 +28,55 @@ func NginxIngressController(cluster metav1.Object) *promv1.ServiceMonitor {
...
+						{
+							TargetLabel: "cluster_type",
+							Replacement: "tenant_cluster",
+						},
+						{
+							TargetLabel: "provider",
+							Replacement: provider,
+						},
+						{
+							SourceLabels: []string{"__meta_kubernetes_service_annotation_giantswarm_io_monitoring"},
```
Comment (mig4) [inline service/controller/resource/servicemonitor/service/nginx-ingress-controller.go:76]:
```suggestion
							SourceLabels: []string{"__meta_kubernetes_service_label_giantswarm_io_monitoring"},
```

File: service/service.go [hunk 11]
```diff
@@ -108,10 +108,13 @@ func New(config Config) (*Service, error) {
...
 		c := clusterapi.ControllerConfig{
-			K8sClient:        k8sClient,
-			Logger:           config.Logger,
-			PrometheusClient: prometheusClient,
-			BaseDomain:       config.Viper.GetString(config.Flag.Service.Prometheus.BaseDomain),
+			K8sClient:                 k8sClient,
+			Logger:                    config.Logger,
+			PrometheusClient:          prometheusClient,
+			BaseDomain:                config.Viper.GetString(config.Flag.Service.Prometheus.BaseDomain),
+			SupportsPersistentStorage: config.Viper.GetBool(config.Flag.Service.Prometheus.Storage.IsPersistent),
```
Comment (mig4) [inline service/service.go:115]: Sorry, I'm a bit lost, what are these for? They don't seem to be used here.

Comment (mig4) [reply service/service.go:115]: Just realised these probably crept-in from #76

File: service/controller/resource/prometheus/resource.go (hunk 3 above)
Comment (omegas27) [reply service/controller/resource/prometheus/resource.go:76]: We have them in g8s-prometheus for https://github.com/prometheus-operator/prometheus-operator/issues/455#issue-239114745

Comment (omegas27) [reply service/controller/resource/prometheus/resource.go:76]: I am not sure what you mean about "add reference to your additionalScrapeConfigs secret here"

File: service/controller/resource/servicemonitor/service/nginx-ingress-controller.go (hunk 9 above)
Comment (omegas27) [reply service/controller/resource/servicemonitor/service/nginx-ingress-controller.go:31]: This is what we have configured in g8s-prometheus and I tried it to check if the networking issue was related. Not sure if I should remove it

File: service/service.go (hunk 11 above)
Comment (omegas27) [reply service/service.go:115]: Yes they did :( I failed to see it

File: service/controller/resource/servicemonitor/service/cluster_autoscaler.go (hunk 5 above)
Comment (omegas27) [reply service/controller/resource/servicemonitor/service/cluster_autoscaler.go:1]: I agree, and I did not plan on making it that big

File: flag/service/prometheus/prometheus.go [hunk 12]
```diff
@@ -2,4 +2,10 @@ package prometheus
 
 type Prometheus struct {
 	BaseDomain string
+	Storage    PrometheusStorage
+}
+
+type PrometheusStorage struct {
+	CreatePVC string
+	Size      string
```
Comment (TheoBrigitte) [inline flag/service/prometheus/prometheus.go:10]: Having the whole storage configuration in a different PR would have been nice.

File: service/controller/resource/prometheus/resource.go [hunk 13]
```diff
@@ -58,6 +64,32 @@ func toPrometheus(v interface{}) (metav1.Object, error) {
 
 	name := cluster.GetName()
 	var replicas int32 = 1
+	var uid int64 = 1000
+	var gid int64 = 65534
+	var fsGroup int64 = 2000
```
Comment (TheoBrigitte) [inline service/controller/resource/prometheus/resource.go:69]: Where do those values come from? A comment would help.

File: files/templates/additional-scrape-configs.template.yaml [hunk 14]
```diff
@@ -0,0 +1,190 @@
+# Additional scrape configuration for the tenant clusters components
+# Add scrape configuration for docker 
+- job_name: {{ .ClusterID }}-prometheus/docker-{{ .ClusterID }}/0
```
[the remark this answers is not in the document]
Comment (omegas27) [reply files/templates/additional-scrape-configs.template.yaml:3]: Joe asked me to have the same names everywhere and those are generated for service monitors

File: flag/service/prometheus/prometheus.go (hunk 12 above)
Comment (omegas27) [reply flag/service/prometheus/prometheus.go:10]: It is in another one. I accidently used the wrong starting branch

Comment (omegas27) [reply flag/service/prometheus/prometheus.go:10]: I will clean it up

File: .idea/.gitignore [hunk 15]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
[the remark this answers is not in the document]
Comment (omegas27) [reply .idea/.gitignore]: Removed :)

File: service/controller/resource/servicemonitor/service/nginx-ingress-controller.go (hunk 9 above)
Comment (omegas27) [reply service/controller/resource/servicemonitor/service/nginx-ingress-controller.go:31]: I removed those values for now

File: files/templates/additional-scrape-configs.template.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (omegas27) [reply files/templates/additional-scrape-configs.template.yaml:3]: I did not see how to rename the generated ones

File: service/controller/resource/prometheus/resource.go (hunk 13 above)
Comment (omegas27) [reply service/controller/resource/prometheus/resource.go]: Done in the other PR

File: files/templates/additional-scrape-configs.template.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (TheoBrigitte) [reply files/templates/additional-scrape-configs.template.yaml:3]: Ok fine, we can do the renaming later.

File: service/controller/resource/prometheus/resource.go (hunk 3 above)
Comment (mig4) [reply service/controller/resource/prometheus/resource.go:111]: Re _reference to your additionalScrapeConfigs_, never mind, I didn't see in the short diff but that's already referenced below, you know the `AdditionalScrapeConfigs: &corev1.SecretKeySelector...`.

Oh so the `ExternalLabels` are labels added to all metrics? TIL 😁
